gooddata-sdk/gooddata_sdk/catalog/user/service.py

```diff
@@ -0,0 +1,97 @@
+# (C) 2022 GoodData Corporation
+from __future__ import annotations
+
+import functools
+from pathlib import Path
+from typing import List, Optional
+
+from gooddata_sdk.catalog.catalog_service_base import CatalogServiceBase
+from gooddata_sdk.catalog.user.declarative_model.user import CatalogDeclarativeUsers
+from gooddata_sdk.catalog.user.declarative_model.user_group import CatalogDeclarativeUserGroups
+from gooddata_sdk.catalog.user.entity_model.user import CatalogUser, CatalogUserDocument
+from gooddata_sdk.catalog.user.entity_model.user_group import CatalogUserGroup, CatalogUserGroupDocument
+from gooddata_sdk.utils import load_all_entities, load_all_entities_dict
+
+
+class CatalogUserService(CatalogServiceBase):
+    def list_users(self) -> List[CatalogUser]:
+        get_users = functools.partial(
+            self._entities_api.get_all_entities_users,
+            include=["userGroups"],
+            _check_return_type=False,
+        )
+        users = load_all_entities_dict(get_users, camel_case=False)
+        return [CatalogUser.from_dict(v, camel_case=False) for v in users["data"]]
+
+    def list_user_groups(self) -> List[CatalogUserGroup]:
+        get_user_groups = functools.partial(
+            self._entities_api.get_all_entities_user_groups,
+            include=["userGroups"],
+            _check_return_type=False,
+        )
+        user_groups = load_all_entities(get_user_groups)
+        return [CatalogUserGroup.from_api(v) for v in user_groups.data]
+
+    def get_user(self, user_id: str) -> CatalogUser:
+        user_dict = self._entities_api.get_entity_users(id=user_id, include=["userGroups"]).data.to_dict(
+            camel_case=False
+        )
+        return CatalogUser.from_dict(user_dict, camel_case=False)
+
+    def create_user(
+        self, user_id: str, authentication_id: Optional[str] = None, user_groups: Optional[list[str]] = None
+    ) -> None:
+        user_document = CatalogUserDocument.create_user(user_id, authentication_id, user_groups)
```

Comment:
This new concept of Document classes and create_x methods is confusing me.
Why is it necessary only in the case of users and user groups?

Reply:
I based that on apidocs where POST /api/entities/users uses object JsonApiUserInDocument and POST /api/entities/userGroups uses object JsonApiUserGroupInDocument. These objects are basically wrappers for  
JsonApiUserIn, JsonApiUserGroupIn. The reason why you might  find this confusing is because these are the first entity objects that are implemented using data classes.